src/wallet/rpc/transactions.cpp

```diff
@@ -875,6 +875,7 @@ RPCHelpMan rescanblockchain()
     uint256 start_block;
 
     EnsureWalletIsUnlocked(*pwallet);
+    LOCK(pwallet->m_relock_mutex);
```

Comment:
In b8952bd7d3b3f745109506223f829c5dd91e638f "wallet: ensure that the passphrase is not deleted from memory when being used to rescan"

The mutex should be grabbed before ensuring the wallet is unlocked. There could be a race condition where the wallet is locked by the scheduler in between checking that the wallet is unlocked, and grabbing `m_relock_mutex`. By grabbing the mutex first, we can be sure that the scheduler will not be able to lock the wallet before we check that the wallet is unlocked.

Also, use `LOCK2` to grab 2 locks at the same time.

Reply:
How would using `LOCK2` work for grabbing `cs_wallet` and `m_relock_mutex` at the same time if the whole point of having `m_relock_mutex` is that we don't want to be holding `cs_wallet` during the time of the rescan since it can be time consuming?

Reply:
`LOCK2` only needs to be used in the places when both locks are being acquired, e.g. in `CWallet::Lock`.

Reply:
I have made it so that `m_relock_mutex` is grabbed before ensuring the wallet is unlocked.
> LOCK2 only needs to be used in the places when both locks are being acquired, e.g. in CWallet::Lock.

Right, I thought you meant this location specifically.